### PR TITLE
Cleanup the bank's use of nonces

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

While addressing a few nonce issues I found the bank's handling of durable nonces has devolved over time using different terms and overly complicated implementations

#### Summary of Changes

- align syntax
- simplify
- general cleanup
- updated comments
- No behavior changes

Fixes #

### CODE REVIEW

**Review by t-nelson — APPROVED** [review]
This looks fine on it's face.  I'm a little uncomfortable with discarding the qualifiers on the durable nonce types and variables since "nonce" is a pretty general concept and if we wind up needing one somewhere else, there's going to be ambiguity.  I'll just assume that you're volunteering to do the re-renames should that situation arise :wink:

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -1002,141 +1000,139 @@ impl Accounts {
...
+                        maybe_nonce,
+                        blockhash,
+                        lamports_per_signature,
+                    );
+
+                    if execution_result.is_ok() || is_nonce_account || is_fee_payer {
+                        if execution_result.is_err() {
+                            if let (false, true, Some(fee_payer_account)) =
+                                (is_nonce_account, is_fee_payer, nonce_fee_payer_account)
+                            {
```
[the remark this answers is not in the document]
Comment (jackcmay) [reply runtime/src/accounts.rs]: Yeah, weird that it was handled separately from the other nonce handling.  I consolidated, like it better now?

**Comment by t-nelson** [comment]
Would it be too much trouble to split the renames out to a separate commit?  I really don't trust myself to review this accurately with them combined

**Comment by jackcmay** [comment]
> Would it be too much trouble to split the renames out to a separate commit? I really don't trust myself to review this accurately with them combined

Sure thing

**Comment by jackcmay** [comment]
@t-nelson this PR now just includes code cleanup, the removal of the special case for nonce instructions, and no blockhash advance if the transaction is processed successfully.  Let me know what you think.

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -1080,102 +1055,99 @@ impl Accounts {
...
-                        .set_state(&NonceVersions::new_current(NonceState::new_initialized(
-                            &data.authority,
-                            blockhash,
-                            lamports_per_signature,
-                        )))
-                        .unwrap();
+            true
+        } else {
+            if
+            /*TODO remove this first part? */
```
Comment (t-nelson) [inline runtime/src/accounts.rs:1136]: As in, do this regardless of the result?

**Review by t-nelson — DISMISSED** [review]
Thanks for pulling out the renames, much easier to read now

File: runtime/src/accounts.rs (hunk 2 above)
Comment (jackcmay) [reply runtime/src/accounts.rs:1136]: Nah, I was playing around with the idea but forget to remove the todo, removed

**Comment by jackcmay** [comment]
@t-nelson I added a feature gone on for the behavior change, can you take another look when you get a chance to see if I captured it correctly?

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -1043,7 +1046,16 @@ impl Accounts {
 
             let (execution_result, nonce) = &execution_results[i];
             let maybe_nonce = match (execution_result, nonce) {
-                (Err(TransactionError::InstructionError(_, _)), Some(nonce)) => Some(nonce),
+                (Ok(_), Some(nonce)) => {
```
Comment (t-nelson) [inline runtime/src/accounts.rs:1049]: Can we either combine this with the other `Ok(_)` match arm or constrain the other one with `None` for the `nonce` field?

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -1047,30 +1047,17 @@ impl Accounts {
...
-                    let address = nonce.address();
-                    let account = nonce.account();
-                    let maybe_fee_payer_account = nonce.fee_payer_account();
-                    Some((
-                        address,
-                        account,
-                        maybe_fee_payer_account,
-                        !nonce_marker_ix_failed,
-                    ))
+                        Some((nonce, false /* rollback */))
```
Comment (t-nelson) [inline runtime/src/accounts.rs:1053]:
```suggestion
                        Some((nonce, false /* don't rollback stored nonce */))
```

It's not clear on first glance that this comment is simply referring to the variable name, so let's just be verbose

File: runtime/src/accounts.rs [hunk 5]
```diff
@@ -1047,30 +1047,17 @@ impl Accounts {
...
+                    }
                 }
-                (Ok(_), _nonce) => None,
-                (Err(_), _nonce) => continue,
+                (Err(TransactionError::InstructionError(_, _)), Some(nonce)) => {
+                    Some((nonce, true /* rollback */))
```
Comment (t-nelson) [inline runtime/src/accounts.rs:1057]:
```suggestion
                    Some((nonce, true /* rollback stored nonce */))
```

**Review by t-nelson — DISMISSED** [review]
lgtm!  I just have a couple clarifying suggestions

File: sdk/src/feature_set.rs [hunk 6]
```diff
@@ -241,6 +241,10 @@ pub mod nonce_must_be_writable {
     solana_sdk::declare_id!("BiCU7M5w8ZCMykVSyhZ7Q3m2SWoR2qrEQ86ERcDX77ME");
 }
 
+pub mod leave_nonce_on_success {
```
Comment (jstarry) [inline sdk/src/feature_set.rs:244]: what does this mean?

**Pull request merged**